lib/backend-wait-logic.js

```diff
@@ -0,0 +1,117 @@
+ * and limitations under the License.
+*/
+"use strict";
+const chalk = require('chalk')
+const ora = require('ora')
+
+const awsExceptionHandler = require('./aws-operations/aws-exception-handler.js')
+const awsClient = require('./aws-operations/aws-client.js')
+const opsProject = require('./backend-operations/ops-project.js')
+const opsCloudApi = require('./backend-operations/ops-cloud-api.js')
+
+const maxCloudApiWaitLoop = 100 //each wait is 5 seconds
+
+function wait(backendDetails, awsDetails, callback){
+    if(isWaitNeeded(backendDetails)){
+        console.log('awsmobile api call returned with no error')
+        console.log('waiting for the mobilehub operations to complete') 
+        let mobile = awsClient.Mobile(awsDetails)
+        let param = {
+            projectId: backendDetails.projectId,
+            syncFromResources: true
+        }
+        cloudApiWaitLoop(mobile, param, backendDetails, 1, callback)
+    }else{
+        callback(null, backendDetails)
+    }
+}
+
+let spinner = ora('waiting ... ') //do not move inside function, the function is called recursively
+function cloudApiWaitLoop(mobile, param, backendDetails, loopCount, callback){
+    spinner.start()
+    if(loopCount > maxCloudApiWaitLoop){
+        console.log(chalk.red('exceeded wait limit'))
+        callback({code: maxCloudApiWaitLoop}, backendDetails)
+    }else{
+        mobile.describeProject(param, function(err,data){
+            spinner.stop()
+            if(err){
+                console.log(chalk.red('wait interrupted'))
+                awsExceptionHandler.handleMobileException(err)
+                callback({code: 0}, backendDetails)
+            }else{
+                backendDetails = data.details
+                if(opsProject.isInNormalState(backendDetails)){
+                    let cloudFormationState = opsCloudApi.getFormationStateSummary(backendDetails)
+                    if(cloudFormationState){
+                        let stateGroup = opsCloudApi.getStateGroup(cloudFormationState)
+                        //status group: 
+                        //-2: unrecognized status
+                        //-1: not yet deployed
+                        // 0: in-progress
+                        // 1: terminal_complate
+                        // 2: terminal_failed
+                        if(stateGroup < -1){
+                            console.log(chalk.red('wait interrupted') + ' unrecognized status code: ' + cloudFormationState)
+                            callback({code: -2}, backendDetails)
+                        }else if(stateGroup == -1){
+                            console.log(chalk.red('wait interrupted') + ' illogical status code: ' + cloudFormationState)
+                            callback({code: -1}, backendDetails)
+                        }else if( stateGroup == 0){
+                            spinner.start('status check #' + chalk.blue(loopCount) + ': ' + cloudFormationState)
```

Comment:
you could have a race condition with the spinner called at the beginning of the method, use loopCount on top instead

Reply:
or stop the spinner before calling the recursive method

Reply:
there's no race condition here, you can call start on a spinner that's already started, there's no problem with that. 
- stop() needs to be called when the loop count reach the max though, and it's added.

Reply:
Cool